multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,377 @@
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if clusterLabels, ok := r.clusterToLabels[clusterID]; ok {
+		clusterLabels.Delete(labelHash)
+	}
+	if clusters, ok := r.labelsToClusters[labelHash]; ok && len(clusters) == 1 && clusters.Has(clusterID) {
+		// The cluster where the label identity is being deleted was the only cluster that has
+		// the label identity. Hence, the label identity is no longer present in the ClusterSet.
+		delete(r.labelsToClusters, labelHash)
+		delete(r.hashToLabels, labelHash)
+		r.labelQueue.Add(labelHash)
+	} else {
+		// Remove mapping from label to cluster
+		clusters.Delete(clusterID)
+	}
+}
+
+// onLabelExportAdd updates the label to cluster caches, and creates LabelIdentity kind of
+// ResourceImport object if needed.
+func (r *LabelIdentityExportReconciler) onLabelExportAdd(clusterID, labelHash, label string) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if labels, ok := r.clusterToLabels[clusterID]; ok {
+		labels.Insert(labelHash)
+	} else {
+		r.clusterToLabels[clusterID] = sets.NewString(labelHash)
+	}
+	if clusters, ok := r.labelsToClusters[labelHash]; !ok {
+		// This is a new label identity in the entire ClusterSet.
+		r.labelsToClusters[labelHash] = sets.NewString(clusterID)
+		r.hashToLabels[labelHash] = label
+		r.labelQueue.Add(labelHash)
+	} else {
+		clusters.Insert(clusterID)
+	}
+}
+
+// Run begins syncing of ResourceImports for label identities.
+func (r *LabelIdentityExportReconciler) Run(stopCh <-chan struct{}) {
+	defer r.labelQueue.ShutDown()
+
+	for i := 0; i < common.DefaultWorkerCount; i++ {
+		go wait.Until(r.labelQueueWorker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (r *LabelIdentityExportReconciler) labelQueueWorker() {
+	for r.processLabelForResourceImport() {
+	}
+}
+
+// Processes an item in the labelQueue. If syncLabelResourceImport returns an error,
+// this function handles it by re-queuing the item so that it can be processed again
+// later. If syncLabelResourceExport is successful, the label is removed from the queue
+// until we get notified of a new change.
+func (r *LabelIdentityExportReconciler) processLabelForResourceImport() bool {
+	key, quit := r.labelQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.labelQueue.Done(key)
+	err := r.syncLabelResourceImport(key.(string))
+	if err != nil {
+		// Put the item back on the workqueue to handle any transient errors.
+		r.labelQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync ResourceImport for label identity", "label", key)
+		return true
+	}
+	// If no error occurs, we forget this item so that it does not get queued again until
+	// another change happens.
+	r.labelQueue.Forget(key)
+	return true
+}
+
+// syncLabelResourceImport checks the label cache and determines whether a ResourceImport
+// needs to be created or deleted for the queued label hash.
+func (r *LabelIdentityExportReconciler) syncLabelResourceImport(labelHash string) error {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
```

Comment:
ditto, does it need to lock the whole method?

Reply:
Fixed